### PR TITLE
Plugins

### PR DESCRIPTION
**Pull request by na9da**
Adds support for loading external plugin components.

The plugin component can be any JS module that exports a [TerriaPlugin](https://github.com/TerriaJS/plugin-sample/blob/83a84f210f56c5d3a2fa73e0417e27cf0fa66312/src/index.ts#L1-L13) definition. It is imported in the plugin.ts file. During app start, terria will load the plugins one by one.

https://github.com/TerriaJS/TerriaMap/blob/8276e4c69201564be381a769249189f636eec1ed/plugins.ts#L1-L6

A [sample plugin repo](https://github.com/TerriaJS/plugin-sample) also has been created.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -75,7 +78,7 @@
     "semver": "^5.0.0",
     "style-loader": "^0.23.1",
     "svg-sprite-loader": "4.1.3",
-    "terriajs": "8.1.25",
+    "terriajs": "https://github.com/terriajs/terriajs#15751503bc6024b7e0cc0cab72f662fd9d32d3a7",
```
Comment (na9da) [inline package.json:81]: To be changed once https://github.com/TerriaJS/terriajs/pull/6212 has been merged

File: package.json [hunk 2]
```diff
@@ -23,7 +25,8 @@
   },
   "dependencies": {
     "pm2": "^3.2.2",
-    "terriajs-server": "^3.3.4"
+    "terriajs-server": "^3.3.4",
+    "terriajs-plugin-api": "https://github.com/terriajs/plugin-api#0.0.1-alpha.1"
```
Comment (na9da) [inline package.json:29]: Change to version tag after publishing in npm.

File: lib/Core/loadPlugins.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+type PluginModule = { default: TerriaPlugin };
+
+/**
+ * Load plugin modules.
+ *
+ * @param pluginPromises Array of promise returning plugin modules.
+ */
+async function loadPlugins(
+  viewState: ViewState,
```
Comment (KeyboardSounds) [inline lib/Core/loadPlugins.ts:15]: Will we ever want to expose more than the ViewState here? If so, I'd make a PluginContext interface like:
```ts
export interface PluginContext {
  viewState: ViewState
}
```

File: lib/Core/loadPlugins.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  pluginPromises: Promise<PluginModule>[]
+): Promise<void> {
+  const loadPromises = pluginPromises.map(promise => {
+    const pluginContext = createPluginContext(viewState);
+    promise
+      .then(({ default: plugin }) => {
+        try {
+          plugin.register(pluginContext);
+        } catch (error) {
+          console.error(`Error when registering plugin "${plugin.name}"`);
```
Comment (KeyboardSounds) [inline lib/Core/loadPlugins.ts:25]: For the sake of consistency, should we be using a TerriaError here?

File: lib/Core/loadPlugins.ts (hunk 3 above)
Comment (na9da) [reply lib/Core/loadPlugins.ts:15]: Yep, that's actually what `createPluginContext` does, it returns a `PluginContext` which is defined in the `plugin-api` repo.

File: lib/Core/loadPlugins.ts (hunk 4 above)
Comment (na9da) [reply lib/Core/loadPlugins.ts:25]: Good point. I'll check how to use it here.

Comment (na9da) [reply lib/Core/loadPlugins.ts]: At this point we don't want to report the error to the user, so a console.error will be enough I think. I have still converted the error to `TerriaError` before logging to console.

**Comment by zoran995** [comment]
Love to see this implemented, @na9da can we just update `plugin-sample` to utilize i18n features although it's a plugin I think that there should be a default way for the internationalization, other than that this will be a great addition

**Comment by na9da** [comment]
Good point @zoran995!

It is actually on my todo list along with figuring out how to package assets.

Also it is currently not possible to build a plugin cleanly independent of terriajs library because terria doesn't emit types. I think plugins will become usable only after we fix https://github.com/TerriaJS/terriajs/issues/5866

**Comment by na9da** [comment]
@KeyboardSounds @steve9164 - I have slightly changed the implementation so that `plugins.ts` now exports a function instead of an array so that the imports are triggered only when the function is called.

**Comment by na9da** [comment]
Hi @steve9164, I have added in the following changes to webpack build process for automatically packaging icon assets from plugin repos.

https://github.com/TerriaJS/TerriaMap/blob/61d2741e280e44ad8c296af8cd0c3c13a1046523/buildprocess/configureWebpackForPlugins.js

**Pull request merged**